multiqc/modules/checkatlas/checkatlas.py

```diff
@@ -0,0 +1,544 @@
+                
+                """,
+            content=table.plot(self.data_metric_dimred),
+        )
+
+    def add_specificitymetrics_section(self):
+        self.add_section(
+            name="Specificity metrics",
+            anchor="checkatlas-specmetrics",
+            description="Quality control metrics calculated on your atlases.",
+            helptext="""
+            
+            """,
+            content="",
+        )
+
+
+def parse_qc_logs(f):
+    """
+    Parse logs from QC tables in .tsv files
+    Order by CellRank
+    Calc log10(CelllRank)
+    :param f:
+    :return:
+    """
+    data_qc_counts = dict()
+    data_qc_genes = dict()
+    data_qc_mito = dict()
+    list_qc_counts = list()
+    list_qc_genes = list()
+    list_qc_mito = list()
+    lines = f.splitlines()
+    headers = lines[0].split("\t")
+    try:
+        index_counts = headers.index(QC_HEADER[0])
+    except ValueError:
+        index_counts = -1
+    try:
+        index_genes = headers.index(QC_HEADER[1])
+    except ValueError:
+        index_genes = -1
+    try:
+        index_mito = headers.index(QC_HEADER[2])
+    except ValueError:
+        index_mito = -1
+
+    for i in range(1, len(lines)):
+        line = lines[i].split("\t")
+        if index_counts != -1:
+            list_qc_counts.append(float(line[index_counts]))
+        if index_genes != -1:
+            list_qc_genes.append(float(line[index_genes]))
+        if index_mito != -1:
+            list_qc_mito.append(float(line[index_mito]))
+    list_qc_counts.sort(reverse=True)
+    list_qc_genes.sort(reverse=True)
+    list_qc_mito.sort(reverse=True)
+
+    for i in range(1, len(list_qc_counts)):
+        data_qc_counts[math.log10(i)] = list_qc_counts[i]
+    for i in range(1, len(list_qc_genes)):
+        data_qc_genes[math.log10(i)] = list_qc_genes[i]
+    for i in range(1, len(list_qc_mito)):
+        data_qc_mito[math.log10(i)] = list_qc_mito[i]
+
+    return data_qc_counts, data_qc_genes, data_qc_mito
+
+
+def parse_firstline_table_logs(f):
+    """
+    Parse only header and first line of logs which are .tsv files
+    Ex: _checkatlas.tsv, adata_checkatlas.tsv
+    :param f:
+    :return:
+    """
+    data = {}
+    lines = f.splitlines()
+    headers = lines[0].split("\t")
+    for i in range(1, len(lines)):
+        line = lines[i].split("\t")
+        for j in range(0, len(line)):
+            data[headers[j]] = line[j]
+    return data
+
+
+def parse_metric_logs(f):
+    """
+    Parse all lines of logs which are .tsv files
+    Ex: _checkatlas.tsv, adata_checkatlas.tsv
+    :param f:
+    :return:
+    """
+    data = {}
+    lines = f.splitlines()
+    headers = lines[0].split("\t")
+    for i in range(1, len(lines)):
+        line = lines[i].split("\t")
+        line_dict = {}
+        for j in range(1, len(line)):
+            line_dict[headers[j]] = line[j]
+        data[line[0]] = line_dict
+    return data
+
+
+def create_img_html_content(type_viz, data):
```

Comment:
Having this kind of code within a single module is very difficult to maintain. See other comment about images more generally, but this is another reason not to use them.

Reply:
Removed